### ISSUE TITLE
lizardfs-uraft metadata server switch to slave mode issue

### ISSUE THREAD

**Issue by jbarrigafk** [lizardfs/lizardfs#818]
hello all, I've been working with lizardfs-uraft 3.13.0 on CentOS 7 and I have problems with my master and shadow servers. I have 3 servers to reach the high availability quorum in the metadata servers. 
Two servers I have in site A and the last in site B, there is communication between the 3 through a VPN site to site, everything works correctly in case one dies, the floating ip changes to another server correctly.
The problem is that after a few days (can be hours or days), the lizardfs-uraft service dies in the master server without any reason, in the logs I have not detected the reason for which the service dies. 
Here I leave the logs and the configuration that I have in my metadata servers. 
I hope you can help me where I should look.

**/var/log/messages**
```
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: Starting metadata server switch to slave mode
Feb 21 02:41:32 mfsmaster2-sv mfsmaster: Can't open PAM session: Operation not permitted
Feb 21 02:41:32 mfsmaster2-sv mfsmaster: can't change open files limit to 5000, check restrictions with ulimit -n
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: 02/21/19 02:41:32.689 [error] [156273:156273] : Can't open PAM session: Operation not permitted
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: 02/21/19 02:41:32.689 [warning] [156273:156273] : can't change open files limit to 5000, check restrictions with ulimit -n
Feb 21 02:41:32 mfsmaster2-sv mfsmaster: changed working directory to: /var/lib/mfs
Feb 21 02:41:32 mfsmaster2-sv mfsmaster: terminate signal received
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:32 mfsmaster2-sv lizardfs-uraft: Node 'mfsmaster1-sv' is now a leader.
Feb 21 02:41:33 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:34 mfsmaster2-sv ntpd[2912]: Deleting interface #7 bond0, x.x.x.100#123, interface stats: received=0, sent=0, dropped=0, active_time=1819545 secs
Feb 21 02:41:34 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:34 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:34 mfsmaster2-sv lizardfs-uraft: Get metadata version timeout.
Feb 21 02:41:37 mfsmaster2-sv lizardfs-uraft: Error: Can't read data from socket: timeout
Feb 21 02:41:37 mfsmaster2-sv lizardfs-uraft: Error: Can't read data from socket: timeout
Feb 21 02:41:38 mfsmaster2-sv lizardfs-uraft: Error: Can't read data from socket: timeout
Feb 21 02:41:38 mfsmaster2-sv lizardfs-uraft: Error: Can't read data from socket: timeout
Feb 21 02:41:42 mfsmaster2-sv mfsmaster: about 10 seconds passed and the lock /var/lib/mfs/.mfsmaster.lock still exists...
Feb 21 02:41:42 mfsmaster2-sv lizardfs-uraft: 02/21/19 02:41:42.601 [warning] [156273:156273] : about 10 seconds passed and the lock /var/lib/mfs/.mfsmaster.lock still exists...
Feb 21 02:41:52 mfsmaster2-sv mfsmaster: about 20 seconds passed and the lock /var/lib/mfs/.mfsmaster.lock still exists...
Feb 21 02:46:22 mfsmaster2-sv lizardfs-uraft: 02/21/19 02:46:22.919 [warning] [156273:156273] : about 290 seconds passed and the lock /var/lib/mfs/.mfsmaster.lock still exists...
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: lockfile /var/lib/mfs/.mfsmaster.lock created and locked
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: sessions have been loaded
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: initialized sessions from file /var/lib/mfs/sessions.mfs
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: initialized exports from file /etc/mfs/mfsexports.cfg
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: initialized topology from file /etc/mfs/mfstopology.cfg
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: initialized goal definitions from file /etc/mfs/mfsgoals.cfg
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: loaded charts data file from /var/lib/mfs/stats.mfs
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: connecting to Master
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: master <-> metaloggers module: listen on *:9419
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: master <-> chunkservers module: listen on *:9420
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: master <-> tapeservers module: listen on (*:9424)
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: main master server module: listen on *:9421
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: open files limit: 1024
Feb 21 02:46:31 mfsmaster2-sv mfsmaster: mfsmaster daemon initialized properly
Feb 21 02:46:31 mfsmaster2-sv mfsmaster[156273]: connected to Master
Feb 21 02:46:31 mfsmaster2-sv lizardfs-uraft: Metadata server switch to slave mode done
Feb 21 02:46:32 mfsmaster2-sv mfsmaster[156273]: terminate signal received
Feb 21 02:46:33 mfsmaster2-sv mfsmaster[156273]: main master server module: closing *:9421
Feb 21 02:46:33 mfsmaster2-sv mfsmaster[156273]: master <-> tapeservers module: closing socket *:9424
Feb 21 02:46:33 mfsmaster2-sv mfsmaster[156273]: master <-> chunkservers module: closing *:9420
Feb 21 02:46:33 mfsmaster2-sv mfsmaster[156273]: master <-> metaloggers module: closing *:9419
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: Metadata server is dead
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: LizardFS uraft helper script
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: Available commands:
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: isalive
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: metadata-version
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: quick-stop
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: promote
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: demote
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: assign-ip
Feb 21 02:46:33 mfsmaster2-sv lizardfs-uraft: drop-ip
```

**lizardfs-uraft.cfg**
mfsmaster2-sv
```
 URAFT_NODE_ADDRESS = mfsmaster2-sv
 URAFT_NODE_ADDRESS = mfsmaster1-sv
 URAFT_NODE_ADDRESS = mfsmaster4-phx
 URAFT_ID = 0
 URAFT_FLOATING_IP      = x.x.x.100
 URAFT_FLOATING_NETMASK = x.x.x.x
 URAFT_FLOATING_IFACE   = bond0
```

mfsmaster1-sv
```
 URAFT_NODE_ADDRESS = mfsmaster2-sv
 URAFT_NODE_ADDRESS = mfsmaster1-sv
 URAFT_NODE_ADDRESS = mfsmaster4-phx
 URAFT_ID = 1
 URAFT_FLOATING_IP      = x.x.x.100
 URAFT_FLOATING_NETMASK = x.x.x.x
 URAFT_FLOATING_IFACE   = bond0
```

mfsmaster4-phx
```
 URAFT_NODE_ADDRESS = mfsmaster2-sv
 URAFT_NODE_ADDRESS = mfsmaster1-sv
 URAFT_NODE_ADDRESS = mfsmaster4-phx
 URAFT_ID = 2
 URAFT_FLOATING_IP      = x.x.x.100
 URAFT_FLOATING_NETMASK = x.x.x.x
 URAFT_FLOATING_IFACE   = bond0
```

**mfsmaster.cfg**
Same config for my 3 metadata servers: mfsmaster1-sv mfsmaster2-sv mfsmaster4-phx
```

 PERSONALITY = ha-cluster-managed
 ADMIN_PASSWORD = somepassword
 WORKING_USER = mfs
 WORKING_GROUP = mfs
 DATA_PATH = /var/lib/mfs
 BACK_LOGS = 25 (I reduced this value to try to reduce the space used in /var/lib/mfs )
 MATOML_LISTEN_HOST = *
 MATOML_LISTEN_PORT = 9419
 MATOCS_LISTEN_HOST = *
 MATOCS_LISTEN_PORT = 9420
 MATOCL_LISTEN_HOST = *
 MATOCL_LISTEN_PORT = 9421
 MATOTS_LISTEN_HOST = *
 MATOTS_LISTEN_PORT = 9424
 NO_ATIME = 0
 MASTER_HOST = mfsmaster
 MASTER_TIMEOUT = 120 (increase this value to see if the lizardfs service dies due to some timeout)

```

When my metadata server dies, my cluster keeps running and I just need to run systemctl start lizardfs-uraft to get back into the cluster, but none should die. Where should I look? If you need more information, please let me know. regards

**Comment by jbarrigafk** [comment]
will have relation with the values ​​of the open files on 10-lizardfs.conf?

**Comment by jbarrigafk** [comment]
im still having issues

`Feb 28 03:00:01 db1-phx systemd: Starting Session 2238 of user root.
Feb 28 03:00:26 db1-phx lizardfs-uraft: Starting metadata server switch to slave mode
Feb 28 03:00:26 db1-phx mfsmaster: Can't open PAM session: Operation not permitted
Feb 28 03:00:26 db1-phx lizardfs-uraft: 02/28/19 03:00:26.908 [error] [21393:21393] : Can't open PAM session: Operation not permitted
Feb 28 03:00:26 db1-phx lizardfs-uraft: 02/28/19 03:00:26.908 [warning] [21393:21393] : can't change open files limit to 5000, check restrictions with ulimit -n
Feb 28 03:00:26 db1-phx mfsmaster: can't change open files limit to 5000, check restrictions with ulimit -n
Feb 28 03:00:26 db1-phx mfsmaster: changed working directory to: /zfs-data/mfs
Feb 28 03:00:26 db1-phx mfsmaster[14448]: terminate signal received
Feb 28 03:00:28 db1-phx ntpd[23893]: Deleting interface #10 bond0, x.x.x.100#123, interface stats: received=0, sent=0, dropped=0, active_time=204540 secs
`

**Comment by jbarrigafk** [comment]
hello all, (I really do not know if someone looks at my problem) but well, doing more tests, I realized that when you use the commands lizardfs-uraft-helper promote or demote, the cluster stops working correctly. These are the tests I did:


**Test1**
start services en all masters (3)
systemctl start lizardfs-uraft
master1 (primary)
master2 (shadow)
master3 (shadow)

on secondary master (shadow, master2)
lizardfs-uraft-helper promote (this command works and promote to master from shadow correctly)

on primary master (master1)
lizardfs-uraft-helper demote ( this command works and demote to shadow from master correctly)

When i reboot my primary master ( master2 now) lizardfs cluster uraft should change any role of the shadow to master, but not works.

**Test2**
I stop everything lizardfs-uraft services on my 3 masters
systemctl stop lizardfs-uraft.
And start again on my 3 masters
systemctl start lizardfs-uraft

master1 (primary)
master2 (shadow)
master3 (shadow)

When i reboot my primary master, lizardfs uraft change roles correctly
master1 (dead)
master2 (primary)
master3 (shadow)

well , returning to my first issue with lizardfs-uraft metadata server switch to slave mode message, doesnt appear at least the last 2 days, new record...
I have no idea if the origin of all the problems is by using the command lizardfs-uraft-helper, because I was testing it since the beginning of this cluster, anyway I will continue monitoring in case the message appears: switch to slave mode.

**Comment by jbarrigafk** [comment]
after a week, no more issues related with lizardfs-uraft metadata server switch to slave mode. I think lizardfs-uraft-helper commands not working at all.

**Comment by phamthanhnhan14** [comment]
Any update for this issue?

**Comment by jbarrigafk** [comment]
hello, I did not have any more problems in HA functionalities with uraft and the change of roles between metadata servers.
The problems appeared because I used the promote and demote commands with lizardfs-uraft-helper.
I did not have problems related to the switch of slave to master or master to slave.

However now I have another problem handling large amounts of files and directories, I opened a new issue for this.
https://github.com/lizardfs/lizardfs/issues/824
